## Problem 1. Suppose we produce ten bootstrapped samples from a data set containing red and green classes. We then apply a classification tree to each bootstrapped sample and, for a specific value of X, produce 10 estimates of P(Class is Red|X): 0.1, 0.15, 0.2, 0.2, 0.55, 0.6, 0.6, 0.65, 0.7, 0.75. There are two common ways to combine these results together into a single class prediction. One is the majority vote approach. The second approach is to classify based on the average probability. In this example, what is the final classification under each of these two approaches?

In [12]:
estimates_p = [0.1, 0.15, 0.2, 0.2, 0.55, 0.6, 0.6, 0.65, 0.7, 0.75]

# Based on the estimate values, let's take a threshold probability as '0.5'.
threshold_p = 0.5

# Approach 1 - Majority Vote
green_count, red_count = 0, 0

for i in estimates_p:
  if i < threshold_p:
    green_count = green_count + 1
  else:
    red_count = red_count + 1

print(red_count, green_count)
print('For majority vote approach final classification is','Green' if red_count < green_count else 'Red')


6 4
For majority vote approach final classification is Red


In [13]:
# Approach 2 - Average Probability
import numpy as np
mean_of_estimates_p = np.mean(estimates_p)
print('Average Probability of class Red is',mean_of_estimates_p)
print('Average Probability of class Green is',1 - mean_of_estimates_p)
print('For average probability approach final classification is', 'Green' if mean_of_estimates_p < threshold_p else 'Red')

Average Probability of class Red is 0.45
Average Probability of class Green is 0.55
For average probability approach final classification is Green



For average probability the P(Red|X) = 0.45 and P(Green|X) = 0.55 which is clearly greater than the threshold ’0.5’, hence final classification for this approach is ’Green’. Whereas in majority vote approach it’s final classification is ’Red’ because the number of estimates for class Red is 6 more when compared to class Green, 4.

Both the approaches produced two different results as their final classifications. The reason over here is that computing mean is sensitive to outliers in the data. As average probability approach uses this, though the number of estimates below the threshold are ’4’, yet they are close to ’0’ rather than to the threshold value, due to this mean value is less than the threshold resulting in a different output

## Problem 2. Provide a detailed explanation of the algorithm that is used to fit a regression tree.


Lets first look at the algorithm thats is used to fit a regression tree.
1. We go for a top to bottom approach and use a recursive splitting on the data
2. We do this by recursively find the best single partitioninig of the data where the reduction of residual sum of squares is highest. This will be a greedy approach.
3. We keep in repeating the above step to each of the split parts individually until we find some minimal number of observations present in each of the leaves
4. Cost complexity pruning is then applied to this larger tree to obtain a sequence of optimal subtrees.

For each value of α there corresponds α subtree T ⊂ T0 such that 

$\sum \limits _{m=i} ^{|T|} \sum \limits _{i:x_{i}∈R_{m}} (y_{i}-\hat{y}_{R_{m})^{2} + \alpha|T|}$

Here |T| represents no. of terminal nodes
The tuning parameter α controls a trade-off between the subtree’s complexity and its fit to the training data

Next we can use K-fold cross-validation to choose α
1. For each value of K we evaluate the mean squared prediction as a function of α on the fold
2. Average the results, and pick α to minimize the average error
3. Taking the α selected in previous step, we return the tree calculated using the formula on the entire dataset with that chosen value of α

Source: 
- https://epurdom.github.io/Stat131A/book/regression-and-classification-trees.html
- https://daviddalpiaz.github.io/stat430fa17/slides/isl/trees.pdf (Pages 17-20)
